### ISSUE TITLE
Slow `deployOnSave`

### ISSUE THREAD

**Issue by jaredmalcolm** [mkloubert/vs-deploy#52]
With `deployOnSave` enabled I see roughly a 9-10 second delay before a single file begins to deploy. 

The actual deployment takes no time at all (less than half a second). If it's needed I'm including my `settings.json` with the specifics removed.

```
{
    "vsicons.presets.angular": false,
    "deploy": {
        "packages":[
            {
                "name": "Package",
                "files": [
                    "**/*.php",
                    "**/*.html",
                    "**/*.htm",
                    "**/*.css",
                    "**/*.js",
                    "**/*.jpg",
                    "**/.htaccess",
                    "**/*.xml",
                    "*.json",
                    "**/*.png"
                ],
                "exclude": [
                    ".vscode/**",
                    ".git/**",
                    ".DS_Store",
                    "**/.DS_Store",
                    ".gitignore",
                    "gulpfile.js",
                    "**/*.scss",
                    "**/*.ts",
                    "node_modules/**",
                    "sftp-config.json"
                ],
                "deployOnSave": true
            }
        ],
        "targets": [{
            "type": "sftp",
            "name": "Live",
            "description": "SFTP content to the Live site",

            "dir": "/home/public_html/",
            "host": "thehost.com",
            "port": 22,
            "user": "username",
            "password": "password"
        }],
        "showPopupOnSuccess": false
    }
}
```

**Comment by mkloubert** [comment]
@jaredmalcolm 

How many files does your configuration matches (about)?

**Comment by jaredmalcolm** [comment]
Do you know of an easy way to determine that? Is the `files` array simply used for the deploy on save or?

**Comment by jaredmalcolm** [comment]
@mkloubert through limiting the values in the `files` array I was able to get the time cut in half. It might be worth adding a note to the docs that as the files that match increase there will be a correlation to the time it takes to detect the changes.

Is there by any chance an event that gets fired in VS Code that would allow you to fire the deploy on save that way?

**Comment by mkloubert** [comment]
You can try to use a tool like [sloc](https://www.npmjs.com/package/sloc), that counts the lines of your code files and also outputs the number of count files.

I am using the [node-glob](https://github.com/isaacs/node-glob) to scan for files inside your workspace.

The workflow, based on your settings, is the following:

1.) scan for `**/*.php` and exclude all files defined in your settings
2.) scan for `**/*.html` and exclude all files defined in your settings
3.) scan for `**/*.htm` and exclude all files defined in your settings
...
10.) scan for `**/*.png` and exclude all files defined in your settings

Because `node-glob` does not support the submission of a file pattern in one single step as array, e.g.

This is why you should try to optimize your patterns.

**Comment by jaredmalcolm** [comment]
@mkloubert Understood! If I select a folder to deploy from the tree. Do any of these rules apply or does it deploy everything or?

**Comment by mkloubert** [comment]
It deploys anything.

**Comment by jaredmalcolm** [comment]
@mkloubert just a thought, but is there a `.onDidSave()` [event](https://code.visualstudio.com/docs/extensionAPI/patterns-and-principles#_events) that might allow you to check a file against a rule instead of interactively checking all the files that match the rules?

**Comment by mkloubert** [comment]
@jaredmalcolm can you give me an example or idea from where you would like to "catch" this event?

**Comment by jaredmalcolm** [comment]
1. File `x` is open in a tab.
2. Changes are made
3 File is saved (`save event is fired by vs code`)
4. Check that file isn't on the exclusion list
5. Deploy

My thought is just that this could provide a near instant check since you are checking that a file isn't excluded and you are only checking a single file instead of all the files that are being tracked.

I don't know what info the VS Code events provide, but I thought this *might* provide a "cheaper" check vs watching everything.

**Comment by jaredmalcolm** [comment]
@mkloubert The `workspace` has an `onDidSaveTextDocument` [event](https://code.visualstudio.com/docs/extensionAPI/vscode-api#_workspace).

Not sure if this provides anything or not.

**Comment by mkloubert** [comment]
@jaredmalcolm 

This is what the extension currently does:

1.) wait for a [onDidSaveTextDocument event](https://github.com/mkloubert/vs-deploy/blob/master/src/extension.ts#L189)
2.) scan for the files, you have defined in the rules / glob patterns
3.) if it [does match](https://github.com/mkloubert/vs-deploy/blob/master/src/deploy.ts#L2249) => deploy the file to the matching targets

I must have a look if these code parts can be optimized or if this could be done by things like an external NodeJS script, which might have a faster way to detect is a file does match a rule for deploying it...

**Comment by mkloubert** [comment]
@jaredmalcolm 

**I have released version 7.25.0 with two new settings `fastCheckOnSave` and `fastCheckOnChange` for [packages](https://github.com/mkloubert/vs-deploy/wiki#packages-).**

If set to `(true)` this will use [minimatch](https://github.com/isaacs/minimatch) instead of [node-glob](https://github.com/isaacs/node-glob), what means that your workspace is not scanned, but simply the path (value) of the underlying file is checked by that module:

```json
{
    "vsicons.presets.angular": false,
    "deploy": {
        "packages":[
            {
                "name": "Package",
                
                "fastCheckOnSave": true
            }
        ],

        // ...
}
```
